### DISCUSSION TITLE
(issue) Publishing transclusions/embeds to sections and blocks

### DISCUSSION THREAD

**Discussion by SemanticBeeng** [ObsidianPublisher/obsidian-github-publisher#190] [Q&A — Plugin] [answered]
Obsiadian can transclude sections of a page by heads or using block references/ids.

![image](https://github.com/ObsidianPublisher/obsidian-github-publisher/assets/9296536/935b8895-a957-4ea0-baf0-d491e9fb2998)

But this Publisher Plugin generates this

![image](https://github.com/ObsidianPublisher/obsidian-github-publisher/assets/9296536/d43fb2f6-3511-4299-8174-e300f4b979f7)

If I remove the `alt` (from the `ApuPpucin` theme to cleanup the padding of the transcluded note) then looks like this

![image](https://github.com/ObsidianPublisher/obsidian-github-publisher/assets/9296536/7cf293c8-cc23-4d16-ae47-565a2f8bd207)
 
 Is this supposed to work?
Missing some setting?

**Comment by Mara-Li** [comment]
Read the documentations, please.
It can't work. Obsidian doesn't have API for the file embedding. I don't want to add it directly in the plugin API. 

You need to use a plugin (for mkdocs), use the settings for removing files embeds, or links.

**Reply by Mara-Li** [reply]
For removing : https://obsidian-publisher.netlify.app/plugin/settings/per%20files%20settings/

**Reply by SemanticBeeng** [reply]
Thanks for clarification.
If Obsidian does not expose transclusions/embeds then it sucks in in unfair kind of way. (forcing their paid Publish plugin).

What you said above: is that said anywhere on the plugin home page in those words exactly?
The link you sent does not make that plainly clear. Read through it before).

> `Obsidian Publisher` does not process transclusions/embeds because Obsidian does not expose apis

It would save others quite a bit of headache to know.

For reference, other plugins that do process transclusions
 * https://github.com/ozntel/ozanshare-publish-plugin (nested)
 * https://publishkit.dev/doc/index.html (maybe)

**Reply by Mara-Li** [reply]
The plugin of ozenshare is pay 
Other plugins doesn't process by their API, but directly with the building website. That why I choose to work on the Mkdocs side instead with the plugin, as it will take a lot of time to :
- Search the embedded files (As last I have with the plugin)
- Search the part to take 
- Take it, but in which formatting? That's the main problem. I can't use HTML, nor set markdown, as users won't be like it in some way.

To be clear (sorry, i'm autistic + English is not my main language), if I understand well how they do the job, the embedding is on the website, not directly with the obsidian API. 
That's why they can add the "flavor" (CSS, html) to set it a way for reader to understand they are reading another file. 

The only proposal I can do is to change the embedding markup (aka `![[]]`) to `→ [[]]` (or `![]()` to `→ []()`)

**Reply by SemanticBeeng** [reply]
Yes, can relate to your thought process.

> " the embedding is on the website, not directly with the obsidian API."

Yup, that is bad, agreed. There should a clear separation of concept between templating/processing of markdown and the rendering (HTML, PDF, etc)

See themes that do nice formatting 
* `AnuPpucin` (https://github.com/AnubisNekhet/AnuPpuccin#simplified-markdown-embeds) and 
* `ebullientworks` (https://github.com/ebullient/obsidian-theme-ebullientworks#style-settings)

But those apply after embeds are resolved.

We need to resolve also heading and and block `[[embed links]]`
 *  `![[embed link#heading]]` 
 *  `![[embed link#^blockId]]` 

> "Take it, but in which formatting? "

Is that the only issue? 
Why not just concat the (markdown) text?

**Reply by Mara-Li** [reply]
For the concating, well, I don't like it. I think it confuse user.
And for the heading and blockID, but normally, it's already taken account by the plugin. I can't add more around that.

**Comment by SemanticBeeng** [comment]
> " for the heading and blockID, but normally, it's already taken account by the plugin"

Forgive me if i misunderstand: from above I get that that the publisher does not process those embeds.

**Reply by Mara-Li** [reply]
Well, maybe I forgot so :'D
I work on that!
I will add new settings to take embed into account, so now you will can transform them into simple links, adding a characters before. Like, `→ [[embed]]` instead of `![[embed]]`

**Reply by Mara-Li** [reply]
btw, you will can change the character!

**Reply by SemanticBeeng** [reply]
Please consider again resolving the embed blocks...
Just extract body of text referenced and include. 
What can go wrong?

**Reply by Mara-Li** [reply]
As I said :
1. It needs too much editing in the file using only regex. 
2. Without a proper citation markup, it can confuse the reader.

**Reply by SemanticBeeng** [reply]
Maybe worth comparing with `Digital Garden` : https://github.com/oleeskild/obsidian-digital-garden/pull/362#issuecomment-1671628939 
It would be great to share a ... canonical implementation.

**Reply by Mara-Li** [reply, marked as answer]
As I said, for the last time.
I don't want to implement directly the block-reference in the file because:
1. I will break my Material Mkdocs template
2. I don't want to confuse user with blockRef, as I think a proper markup is needed (and `>` is not good).

You can : 
- Implement and test it yourself with a fork of my plugin, maybe with an option. 
- Use my Mkdocs template
- Use any template that supports transclusion
- Use Digital Garden and see the result.

**Reply by ghost** [reply]
@SemanticBeeng  sorry for chiming in, but it's a free project maintained by someone on their free time, please be respectful and don't act like they owe you something :)

**Reply by SemanticBeeng** [reply]
No, the world and me owe you something for this great work! Thank you!

Sorry, i'm a bit autistic myself ... and am seeing partial solutions in a number of places. 
Looking to understand, clarify and improve. Even to make a small donation if it helps.

**Comment by SemanticBeeng** [comment]
> * Use my Mkdocs template
> * Use any template that supports transclusion

Forgive my ignorance.
Is this Mkdocs template supposed to handle the `[[embed links]]` exactly like in Obsidian client (as per https://github.com/orgs/ObsidianPublisher/discussions/190#discussioncomment-6679194) ?

In other words you are saying that should be possible to achieve this functionality using `MkDocs`?
Looking at `snippets` to learn https://squidfunk.github.io/mkdocs-material/setup/extensions/python-markdown-extensions/?h=snippets#snippets

See related https://github.com/mokeyish/obsidian-enhancing-export/issues/74 about Obsidian roadmap on these issues.

**Reply by Mara-Li** [reply]
Did you read the documentation of the plugin?

I worked on a Mkdocs template, with a plugin that handle embed...

**Reply by SemanticBeeng** [reply]
Read all under https://obsidian-publisher.netlify.app/ looking for specific examples or mentions about `![[embed link#heading]]` and `![[embed link#^blockId]]` and could not find any so far. 

Will keep looking in code trying to make sense how then Mkdocs template achieves same rendering functionality as in Obsidian client.

**Reply by Mara-Li** [reply]
Oh, well.

Maybe I need to add the code for this page : https://obsidian-publisher.netlify.app/template/example/

I will do it later.

**Reply by SemanticBeeng** [reply]
Indeed, not clear at this time.
It would help a lot to show relation to `Obsidian` client rendering for all kinds of embeds as well.

**Reply by Mara-Li** [reply]
???? My plugin doesn't do anything to Obsidian file. Embedding depend of your theme ??
It's litteraly the same as in Obsidian…

**Comment by Mara-Li** [comment]
Hello
With the publishing of "easy-bake", I take a look into the code and added it into "obsidian-publisher" with proper credit 
See [bake-embed](https://github.com/ObsidianPublisher/obsidian-github-publisher/blob/bake-embed/src/conversion/bakeEmbed.ts)

**Comment by SemanticBeeng** [comment]
Just an fyi

At this time, with latest plugin, getting this effect for embeds

In `Obsidian`
![image](https://github.com/ObsidianPublisher/obsidian-github-publisher/assets/9296536/19a77a82-bad3-41a9-a2ed-f68ccc356644)

In generated git repo

![image](https://github.com/ObsidianPublisher/obsidian-github-publisher/assets/9296536/ef311772-8212-4c2b-a0b5-6d415a3f0ea3)

---
Also
![image](https://github.com/ObsidianPublisher/obsidian-github-publisher/assets/9296536/d89b5c3b-47d1-4d85-b58d-3946ee8d8cba)

vs
![image](https://github.com/ObsidianPublisher/obsidian-github-publisher/assets/9296536/e0aeec2c-8c0f-4b9a-8241-5641ea14f029)


Forgive me if still missing the point.

**Reply by Mara-Li** [reply]
If you use the latest beta just check the embed settings, please.

Also beware, a lot is in beta. I try to refactoring the code but, for an unknown reason, Webstorm and VScode intellisense loop crash. I need more RAM...

**Comment by SemanticBeeng** [comment]
Updated the plugin in Obsidian. 
Cannot see a version so unsure if "beta"

![image](https://github.com/ObsidianPublisher/obsidian-github-publisher/assets/9296536/d6c08d85-7a99-41a0-bd38-217586df621c)

![image](https://github.com/ObsidianPublisher/obsidian-github-publisher/assets/9296536/e661f159-c982-4488-b78f-311b4c89e345)

Which settings should I see?

**Reply by Mara-Li** [reply]
You need to use brat... It's in the documentation...
